commons/src/lexer/token/iterator/mod.rs

```diff
@@ -0,0 +1,1284 @@
+    /// Pushes the new iterator to a new scope, and only runs the given matching functions in the new scope.
+    ///
+    /// # Arguments
+    ///
+    /// * `prefix_match` ... Optional matching function used to strip prefix on new lines
+    /// * `end_match` ... Optional matching function used to indicate the end of the created iterator
+    pub fn nest_scoped(
+        mut self,
+        prefix_match: Option<IteratorPrefixFn>,
+        end_match: Option<IteratorEndFn>,
+    ) -> TokenIterator<'slice, 'input> {
+        let start_index = self.index();
+        let iter_end = self.iter_end;
+        let prefix_mismatch = self.prefix_mismatch;
+        let prefix_start = self.prefix_start;
+
+        let scope = self.scope + 1;
+        self.set_curr_scope(scope);
+
+        TokenIterator {
+            parent: TokenIteratorKind::Nested(Box::new(self)),
+            start_index,
+            match_index: 0,
+            scope,
+            scoped: true,
+            skip_end_until_idx: 0,
+            prefix_match,
+            end_match,
+            iter_end,
+            prefix_mismatch,
+            next_matching: false,
+            peek_matching: false,
+            prefix_start,
+        }
+    }
+
+    /// Nests this iterator, by creating a new iterator that has this iterator set as parent.
+    /// Pushes the new iterator to a new scope root, and only runs the given matching functions in the new scope.
+    /// Parent iterators do not get updated to the new scope, so matching functions for scoped parent iterators are still executed.
+    ///
+    /// # Arguments
+    ///
+    /// * `prefix_match` ... Optional matching function used to strip prefix on new lines
+    /// * `end_match` ... Optional matching function used to indicate the end of the created iterator
+    pub fn new_scope_root(
+        self,
+        prefix_match: Option<IteratorPrefixFn>,
+        end_match: Option<IteratorEndFn>,
+    ) -> TokenIterator<'slice, 'input> {
+        let start_index = self.index();
+
+        TokenIterator {
+            parent: TokenIteratorKind::ScopedRoot(TokenIteratorScopedRoot::from(self)),
+            scope: 0,
+            scoped: false,
+            skip_end_until_idx: 0,
+            start_index,
+            match_index: 0,
+            prefix_match,
+            end_match,
+            iter_end: false,
+            prefix_mismatch: false,
+            next_matching: false,
+            peek_matching: false,
+            prefix_start: None,
+        }
+    }
+
+    /// Returns `true` if this iterator is nested.
+    pub fn is_nested(&self) -> bool {
+        matches!(
+            self.parent,
+            TokenIteratorKind::Nested(_) | TokenIteratorKind::ScopedRoot(_)
+        )
+    }
+
+    /// Returns `true` if this iterator created a new scope.
+    pub fn is_scoped(&self) -> bool {
+        self.scoped
+    }
+
+    /// Returns the parent iterator if this iterator is nested.
+    pub fn unfold(self) -> Self {
```

Comment:
If we change `parent` to `inner`, this function might be renamed to `into_inner` or just `inner` as well.